src/PhpDocument.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace LanguageServer;
+
+use \LanguageServer\Protocol\{Diagnostic, DiagnosticSeverity, Range, Position, SymbolKind, TextEdit};
+
+use PhpParser\{Error, Comment, Node, ParserFactory, NodeTraverser, Lexer, Parser};
+use PhpParser\PrettyPrinter\Standard as PrettyPrinter;
+use PhpParser\NodeVisitor\NameResolver;
+
+class PhpDocument
+{
+    private $stmts;
+    private $client;
+    private $project;
+    private $symbols = [];
+    private $parser;
+    private $uri;
+
+    public function __construct(string $uri, Project $project, LanguageClient $client, Parser $parser)
+    {
+        $this->uri = $uri;
+        $this->project = $project;
+        $this->client = $client;        
+        $this->parser = $parser;
+    }
+
+    /**
+     * Returns all symbols in this document.
+     *
+     * @return SymbolInformation[]
+     */
+    public function getSymbols()
+    {
+        return $this->symbols;
+    }
+
+    /**
+     * Returns symbols in this document filtered by query string.
+     *
+     * @param string $query The search query
+     * @return SymbolInformation[]
+     */
+    public function findSymbols(string $query)
+    {
+        return array_filter($this->symbols, function($symbol) use(&$query) {
+            return stripos($symbol->name, $query) !== false;

```

Comment:
I would not implement it myself but look for a library from composer that we can integrate (or improve so it can be integrated here).